core/src/smartcontracts/isi/asset.rs

```diff
@@ -54,89 +54,165 @@ pub mod isi {
-            let asset_id = self.destination_id;
+    impl FloatMetricFrom for u128 {
+        #[allow(clippy::cast_precision_loss)]
+        fn from(value: Self) -> f64 {
+            value as f64
+        }
+    }
 
-            assert_can_mint(&asset_id.definition_id, wsv, AssetValueType::Quantity)?;
-            wsv.asset_or_insert(&asset_id, 0_u32)?;
-            wsv.modify_asset(&asset_id, |asset| {
-                let quantity: &mut u32 = asset
-                    .try_as_mut()
-                    .map_err(eyre::Error::from)
-                    .map_err(|e| Error::Conversion(e.to_string()))?;
-                *quantity = quantity
-                    .checked_add(self.object)
-                    .ok_or(MathError::Overflow)?;
-                wsv.metrics.tx_amounts.observe(f64::from(*quantity));
+    impl FloatMetricFrom for u32 {
+        fn from(value: Self) -> f64 {
+            value.into()
+        }
+    }
 
-                Ok(AssetEvent::Added(asset_id.clone()))
-            })
+    impl FloatMetricFrom for Fixed {
+        fn from(value: Self) -> f64 {
+            value.into()
         }
     }
 
-    impl<W: WorldTrait> Execute<W> for Mint<Asset, u128> {
-        type Error = Error;
+    /// Trait for `try_checked_add` operation on an asset value that can fail and return [`Error`].
+    trait TryCheckedAdd {
+        fn try_checked_add(self, rhs: Self) -> Result<Self, Error>
+        where
+            Self: Sized;
+    }
 
-        #[metrics(+"mint_big_qty")]
-        fn execute(
-            self,
-            _authority: <Account as Identifiable>::Id,
-            wsv: &WorldStateView<W>,
-        ) -> Result<(), Self::Error> {
-            let asset_id = self.destination_id;
+    impl TryCheckedAdd for u32 {
+        fn try_checked_add(self, rhs: Self) -> Result<Self, Error> {
+            Ok(self.checked_add(rhs).ok_or(MathError::Overflow)?)
+        }
+    }
 
-            assert_can_mint(&asset_id.definition_id, wsv, AssetValueType::BigQuantity)?;
-            wsv.asset_or_insert(&asset_id, 0_u128)?;
-            wsv.modify_asset(&asset_id, |asset| {
-                let quantity: &mut u128 = asset
-                    .try_as_mut()
-                    .map_err(eyre::Error::from)
-                    .map_err(|e| Error::Conversion(e.to_string()))?;
-                *quantity = quantity
-                    .checked_add(self.object)
-                    .ok_or(MathError::Overflow)?;
-                #[allow(clippy::cast_precision_loss)]
-                wsv.metrics.tx_amounts.observe(*quantity as f64);
+    impl TryCheckedAdd for u128 {
+        fn try_checked_add(self, rhs: Self) -> Result<Self, Error> {
+            Ok(self.checked_add(rhs).ok_or(MathError::Overflow)?)
+        }
+    }
 
-                Ok(AssetEvent::Added(asset_id.clone()))
-            })
+    impl TryCheckedAdd for Fixed {
+        fn try_checked_add(self, rhs: Self) -> Result<Self, Error> {
+            Ok(self.checked_add(rhs)?)
```

Comment:
I think we can unify this error handling with other cases. Then the new trait would not be necessary

```suggestion
            Ok(self.checked_add(rhs).ok_or(MathError::Overflow)?)
```

Reply:
Problem is that `Fixed::checked_add` returns `Result`, not `Option` like `u128::checked_add` or `u32::checked_add` and `ok_or` does not exists for `Result`

Reply:
That can be changed, to make the API for `Fixed` more in line with other numerical types.

Reply:
But we still need to define `Checked(Add|Sub)` traits because we are not using crates like `num` or `num_taits` where they are defined for us (and Fixed is our custom type anyway), right?  Or i miss something?

I mean:

```rust
/// Trait for `checked_add` operation on iroha's primitive types.
pub trait CheckedAdd {
    fn checked_add(self, rhs: Self) -> Option<Self>
    where
        Self: Sized;
}

// impls for u32, u128

impl CheckedAdd for Fixed {
    #[inline]
    fn checked_add(self, rhs: Self) -> Option<Self> {
        self.checked_add(rhs).ok()
    }
}

fn test(a: impl CheckedAdd) -> Result<(), MathError> {
    // ...
    let c = a.checked_add(b).ok_or(MathError::Overflow)?
    Ok(())
}

```